src/picture.cpp

```diff
@@ -503,40 +511,49 @@ static surface load_image_data_uri(const image::locator& loc)
 	return surf;
 }
 
-// small utility function to store an int from (-256,254) to an signed char
-static signed char col_to_uchar(int i)
+light_adjust::light_adjust(int op, int rr, int gg, int bb)
+	: l(op), r(), g(), b()
 {
-	return static_cast<signed char>(std::min<int>(127, std::max<int>(-128, i / 2)));
+	constexpr int min = std::numeric_limits<int8_t>::min();
+	constexpr int max = std::numeric_limits<int8_t>::max();
+
+	// Clamp in a wider type (int) to avoid truncating the input value prematurely
+	r = std::clamp(rr / 2, min, max);
+	g = std::clamp(gg / 2, min, max);
+	b = std::clamp(bb / 2, min, max);
 }
 
-light_string get_light_string(int op, int r, int g, int b)
+namespace
+{
+std::size_t hash_light_range(const utils::span<const light_adjust>& range)
 {
-	light_string ls;
-	ls.reserve(4);
-	ls.push_back(op);
-	ls.push_back(col_to_uchar(r));
-	ls.push_back(col_to_uchar(g));
-	ls.push_back(col_to_uchar(b));
-
-	return ls;
+	std::size_t hash{0};
+	for(const auto& adjustment : range) {
+		hash += adjustment.l + adjustment.r + adjustment.g + adjustment.b;
```

Comment:
since l,r,g,h are all int8s this seems like a very bad hash function, that never uses the full range of the target type (size_t)

Reply:
even if you do the static cast the hash function seemd bad, since adding a few number smaller than 255 will still nerever use the full range of a size_t


also for exmaple switching the `r` and `g` leads to the same hash, I'm pretty sure that the game has cases where we have multiple [times] the just differ by switching rgb values.

Reply:
Ah, I misunderstood your comment. I thought you meant that adding int8_t thrice would incorrectly result in overflow. You're right, though, size_t is too large in this case. I'll reduce it to an int.

Reply:
@gfgtdf could you have a look at #10154?